libx{dmcp,au,cb,11,ext}: refactor and move to pkgs/by-name from xorg namespace

quantenzitrone opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

After this pr is merged i will be able to make several small chunk PRs that can be merged independently without merge conflicts.

my plan is to migrate all packages from xorg to pkgs/by-name
it looks like chunks of 5 packages seem ok small chunks

this is the script i use to get the packages that don't depend on other xorg packages:

```nix
let
  inherit (import <nixpkgs> { }) lib;

  done = import ./pkgs/servers/x11/xorg/default.nix |> lib.functionArgs |> lib.attrNames;

  pkgs =
    (import ./default.nix { }).xorg
    |> lib.filterAttrs (
      n: v:
      lib.isAttrs v
      && v ? pname
      && !lib.elem n done
      && !lib.elem n (map (lib.replaceStrings [ "-" ] [ "" ]) done)
      && !lib.elem v.pname done
      && !lib.elem v.pname (map (lib.replaceStrings [ "-" ] [ "" ]) done)
      && !lib.elem (lib.toLower n) done
      && !lib.elem (lib.toLower n) (map (lib.replaceStrings [ "-" ] [ "" ]) done)
      && !v.meta.unfree
      && !v.meta.broken
      && !v.meta.unsupported
    );

  pnames = lib.mapAttrsToList (_: v: v.pname) pkgs;
in
pkgs
# get deps
|> lib.mapAttrs (
  _: v:
  v.buildInputs ++ v.nativeBuildInputs ++ v.propagatedBuildInputs ++ v.propagatedNativeBuildInputs
)
# filter deps for xorg packages
|> lib.mapAttrs (_: v: lib.filter (x: x ? pname && lib.elem x.pname pnames) v)
# map deps to list and count
|> lib.mapAttrsToList (
  name: value: {
    inherit name;
    deps = map (x: x.pname or "error") value |> lib.unique;
    count = lib.length value;
  }
)
# sort
|> lib.sort (x: y: x.count > y.count)
# represent as string
|> map (x: "${toString x.count} - ${toString x.name}: ${lib.concatStringsSep " " x.deps}")
```
you can call it like this to get a nice output:
```sh
nix eval --json --file ./xorg-nodepends.nix | jq .[] --raw-output
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

quantenzitrone committed:
libxdmcp: refactor, move to pkgs/by-name and rename from xorg.libXdmcp

quantenzitrone committed:
libxau: refactor, move to pkgs/by-name and rename from xorg.libXau

quantenzitrone committed:
libxcb: refactor and move to pkgs/by-name from xorg namespace

this removes all the configure flags because they are either default in
stdenv (--disable-shared on static) or are enabled by default and thus
not needed anymore

this also removes libxslt from the builtInputs since it seems to not be
needed according to nix-check-deps (if this makes problems please re-add it)

quantenzitrone committed:
libxext: refactor, move to pkgs/by-name and rename from xorg.libXext

quantenzitrone commented:
i went through the packages with nix-check-deps and found that libxcb and libx11 don't need libpthread-stubs and libxslt

alyssais commented on pkgs/by-name/li/libx11/package.nix:
```diff
@@ -0,0 +1,100 @@
+
+  depsBuildBuild = [
+    buildPackages.stdenv.cc
+  ] ++ lib.optional stdenv.hostPlatform.isStatic (buildPackages.stdenv.cc.libc.static or null);
```
This seems odd to me, because we're checking the host platform but then adding inputs for the build platform. I know it was previously xorg.buildPackages.stdenv, so maybe buildPackages.stdenv is something else? Have you tried a static build? I'd expect it to be 

```suggestion
  ] ++ lib.optional stdenv.hostPlatform.isStatic (stdenv.cc.libc.static or null);
```

quantenzitrone replied:
Wouldn't `depsBuildBuild` have to come from `pkgsBuildBuild` and not `buildPackages` which according to the nixpkgs-manual is an alias to `pkgsBuildHost`?

quantenzitrone replied:
this just gave me a major brainfuck so i asked on discourse https://discourse.nixos.org/t/why-is-buildpackages-aka-pkgsbuildhost-used-in-depsbuildbuild-instead-of-pkgsbuildbuild/65334

quantenzitrone replied:
found it `nix build .#pkgsStatic.libx11`

quantenzitrone replied:
since `xorg.buildPackages` doesn't exist, `xorg.buildPackages.libc.static or null` evaluates to `null`
so the the line didn't do anything

adding `libc.static` actually breaks the static build in a way i can't explain

quantenzitrone committed:
libx11: refactor, move to pkgs/by-name and rename from xorg.libX11

refactorial changes:
- move `CPP` into `env`
- remove `sed 's,^as_dummy.*,as_dummy="\$PATH",' -i configure` because
  it didn't change anything other than the store paths in the resulting
  derivation (using `diff --recursive` (and `xxd` to diff binary files
  that changed))
- remove `xorg.buildPackages.libc.static` from `depsBuildBuild` since
  `xorg.buildPackages` doesn't exist (thus
  `xorg.buildPackages.libc.static or null` evaluating to `null`).
  Fixing that to `libc.static` (and moving it to `nativeBuildInputs`
  (where non-compiler deps belong)) actually broke the static build in
  a weird way i don't understand.

quantenzitrone commented:
I will have to wait until #414573 is in at least nixpkgs-unstable-small because it creates a merge conflict with my libx11 migration changes and thus breaks my workflow consisting of cherry-picking my commits on nixos-unstable to build and test them.

I can't build staging on my PC and i also prefer not to build from master.

quantenzitrone commented:
The builds:
```
nom build .#{,pkgsStatic.,pkgsCross.aarch64-multiplatform{,-musl}.}libx{dmcp,au,cb,11,ext}
```
when cherry-picked on a bit outdated master.

All succeed on my `x86_64-linux` system.

Ericson2314 commented:
Will we loose the ability to generate things automatically with this?

quantenzitrone commented:
i think i still prefer the first option, as automated independent updates are better than manual ones.
Also I'm kinda too far into the moving process now, to decide to instead change the `update.py`.

alyssais commented on pkgs/by-name/li/libxcb/package.nix:
```diff
@@ -0,0 +1,96 @@
+  ];
+  propagatedBuildInputs = lib.optional stdenv.hostPlatform.isMinGW windows.mingw_w64_pthreads;
+
+  # $dev/include/xcb/xcb.h includes pthread.h
```
This comment is now in the wrong place. It should be attached to propagatedBuildInputs.

quantenzitrone replied:
not sure how that happened, thanks for catching it

alyssais commented on pkgs/by-name/li/libxcb/package.nix:
```diff
@@ -0,0 +1,96 @@
+  configureFlags = [
+    "--enable-xkb"
+    "--enable-xinput"
+  ] ++ lib.optional stdenv.hostPlatform.isStatic "--disable-shared";
```
You could drop this if you want — stdenv does it automatically.

quantenzitrone replied:
i also found out that the other 2 configureFlags do nothing

quantenzitrone commented:
i need to re-add libpthread-stubs, but i don't have the time rn so to prevent merge i marked it as draft

alyssais commented:
I'd love to get this merged when you have the time :)

quantenzitrone commented:
sorry, i procastinated wayy to much
not only on this also on my exams
but i'm done with then for now :)

alyssais commented on pkgs/by-name/li/libx11/package.nix:
```diff
@@ -0,0 +1,94 @@
+
+  strictDeps = true;
+
+  depsBuildBuild = [ pkgsBuildHost.stdenv.cc ];
```
Can we leave this as buildPackages?  It's just much more common for us to write `buildPackages.stdenv.cc` than `pkgsBuildHost.stdenv.cc` — the latter looks a bit weird juxtaposed with `depsBuildBuild` even though it's correct.

quantenzitrone replied:
I had a huge confusion, because normally this could just be `stdenv.cc` due to splicing selecting the correct attribute.
However this is not the case because `stdenv.cc` is a dependency of `stdenv`, so it is "one up" meaning `pkgsBuildHost.stdenv.cc` is hosted on the buildPlatform of libx11s buildPlatform and targeted on libx11s buildPlatform,  aka an offset of (-2, -1).
Since we are not bootstrapping here, this is the same as `pkgsBuildBuild` (offset (-1,-1)).
At least that is how I understand it. Maybe that is not fully accurate.
Emily had to explain this to me on matrix very lengthily.

I guess we can use `buildPackages`, but i think these names are terribly confusing since they don't accurately portray what is going on.

quantenzitrone replied:
alright

alyssais commented on pkgs/by-name/li/libx11/package.nix:
```diff
@@ -0,0 +1,94 @@
+
+  postInstall = ''
+    # Remove useless DocBook XML files.
+    test -d $out/share/doc # fail if $out/share/doc doesn't exist
```
Could we avoid this by dropping `-f` from rm?

trofi commented:
Bisect says 3c1d6186b099969e1f9817dd9ca0711b61a69a79 `libx11: refactor, move to pkgs/by-name and rename from xorg.libX11` broke `libva-vdpau-driver` in `staging` as:

```
$ nix build --impure --expr 'with import ./. { overlays = [(final: prev: { libarchive = prev.libarchive.overrideAttrs (oa: { doCheck = false; }); })]; }; libva-vdpau-driver' -L
...
libva-vdpau-driver> checking for LIBVA_DEPS... yes
libva-vdpau-driver> checking for LIBVA_X11_DEPS... no
libva-vdpau-driver> configure: error: Package requirements (libva-x11 >= 0.31.0) were not met:
libva-vdpau-driver>
libva-vdpau-driver> No package 'libva-x11' found
libva-vdpau-driver>
libva-vdpau-driver> Consider adjusting the PKG_CONFIG_PATH environment variable if you
libva-vdpau-driver> installed software in a non-standard prefix.
libva-vdpau-driver>
libva-vdpau-driver> Alternatively, you may set the environment variables LIBVA_X11_DEPS_CFLAGS
libva-vdpau-driver> and LIBVA_X11_DEPS_LIBS to avoid the need to call pkg-config.
libva-vdpau-driver> See the pkg-config man page for more details.
error: Cannot build '/nix/store/2071ninraf9p6kzjxhr98zzby8nvcdzw-libva-vdpau-driver-0.7.4.drv'.
```

quantenzitrone commented:
yeah please revert it
i'll look into it, but it will probably take some time

quantenzitrone commented:
the github revert button does not work, has it been reverted yet?

quantenzitrone commented:
it is really weird, because in the build logs of libva it says
```
Run-time dependency x11-xcb found: NO (tried pkgconfig and cmake)
Run-time dependency xcb found: NO (tried pkgconfig and cmake)
Run-time dependency xcb-dri3 found: NO (tried pkgconfig and cmake)
```

however libx11 clearly has an `$out/lib/pkg-config/x11-xcb.pc` file
